model/common/src/icon4py/model/common/orchestration/decorator.py

```diff
@@ -0,0 +1,673 @@
+        else:
+            my_rank = 0 if not mpi4py else MPI.COMM_WORLD.Get_rank()
+            mpi_size = 1 if not mpi4py else MPI.COMM_WORLD.Get_size()
+        l_uids.append(uid_from_hashlib(str({"MPI_rank": my_rank, "MPI_size": mpi_size})))
+
+        unique_id = uid_from_hashlib(str(l_uids))
+
+    return unique_id
+
+
+def generate_orchestration_uid(
+    obj: Any, obj_name: str = "", members_to_disregard: tuple[str] = ()
+) -> str:
+    """Generate a unique id for a runtime object.
+
+    The unique id is generated by creating a dictionary that describes the runtime state of the object.
+    For primitive types, the dictionary contains the type and the value.
+    For arrays, the dictionary contains the shape and the dtype -not their content-.
+
+    Keep in mind that this function is not supposed to be generic, but it is used only for the DaCe orchestration purposes.
+    """
+    primitive_dtypes = (*orchestration_dtypes.ICON4PY_PRIMITIVE_DTYPES, str, uuid.UUID, np.dtype)
+
+    unique_dict = {}
+
+    def _populate_entry(key: str, value: Any, parent_key: str = ""):
+        full_key = f"{parent_key}.{key}" if parent_key else key
+
+        if full_key in members_to_disregard:
+            return
+
+        if isinstance(value, primitive_dtypes):
+            unique_dict[full_key] = {"type": "primitive_dtypes", "value": str(value)}
+        elif isinstance(value, (np.ndarray, gtx.Field)):
+            unique_dict[full_key] = {
+                "type": "array/field",
+                "shape": str(value.shape),
+                "dtype": str(value.dtype),
+            }
+        elif isinstance(value, (list, tuple)):
+            if all(isinstance(i, primitive_dtypes) for i in value):
+                unique_dict[full_key] = {
+                    "type": f"array-like[{'empty' if len(value) == 0 else type(value[0])}]",
+                    "length": str(len(value)),
+                }
+            else:
+                for i, v in enumerate(value):
+                    _populate_entry(str(i), v, full_key)
+        elif value is None:
+            unique_dict[full_key] = {"type": "None", "value": "None"}
+        elif hasattr(value, "__dict__") or isinstance(value, dict):
+            _populate_unique_dict(value, full_key)
+        else:
+            raise ValueError(f"Type {type(value)} is not supported.")
+
+    def _populate_unique_dict(obj: Any, parent_key: str = ""):
+        if (hasattr(obj, "__dict__") or isinstance(obj, dict)) and not isinstance(
+            obj, decomposition.ExchangeRuntime
+        ):
+            obj_to_traverse = obj.__dict__ if hasattr(obj, "__dict__") else obj
+            for key, value in obj_to_traverse.items():
+                _populate_entry(key, value, parent_key)
+
+    if hasattr(obj, "__dict__") or isinstance(obj, dict):
+        _populate_unique_dict(obj)
+    else:
+        _populate_entry(obj_name, obj)
+
+    return uid_from_hashlib(str(unique_dict))
```

Comment:
The names of this function and the previous one do not follow a similar naming scheme despite they are closely related, I would try to rename them. Additionally, both functions are lacking the arg description in their docstrings and this one seems more convoluted than needed. I would suggest to refactor it to something with a single internal helper function like:

```suggestion
def generate_orchestration_uid(
    obj: Any, obj_name: str = "", members_to_disregard: tuple[str] = ()
) -> str:
    """Generate a unique id for a runtime object.

    The unique id is generated by creating a dictionary that describes the runtime state of the object.
    For primitive types, the dictionary contains the type and the value.
    For arrays, the dictionary contains the shape and the dtype -not their content-.

    Keep in mind that this function is not supposed to be generic, and should only be used for
    DaCe orchestration purposes.

    Args:
        obj:
        obj_name:
        members_to_disregard:
    """
    primitive_dtypes = (*orchestration_dtypes.ICON4PY_PRIMITIVE_DTYPES, str, uuid.UUID, np.dtype)

    static_data = {}

    def _populate_entry(key: str, value: Any) -> None:
        if key in members_to_disregard:
            return

        if isinstance(value, primitive_dtypes):
            static_data[key] = {"type": "primitive_dtypes", "value": str(value)}

        elif isinstance(value, (np.ndarray, gtx.Field)):
            static_data[key] = {
                "type": "array/field",
                "shape": str(value.shape),
                "dtype": str(value.dtype),
            }

        elif isinstance(value, (list, tuple)):
            if all(isinstance(i, primitive_dtypes) for i in value):
                static_data[key] = {
                    "type": f"array-like[{'empty' if len(value) == 0 else type(value[0])}]",
                    "length": str(len(value)),
                }
            else:
                for child_key, child_value in enumerate(value):
                    _populate_entry(f"{key}.{child_key!s}", child_value)

        elif isinstance(value, decomposition.ExchangeRuntime):
            pass

        elif value is None:
            static_data[key] = {"type": "None", "value": "None"}

        elif isinstance(value, dict) or getattr(obj, "__dict__", None):
            for child_key, child_value in getattr(obj, "__dict__", obj).items():
                _populate_entry(f"{key}.{child_key!s}", child_value)

        else:
            raise ValueError(f"Type {type(value)} is not supported.")

    _populate_entry(obj_name, obj)

    return uid_from_hashlib(str(static_data))
```

Reply:
I have updated the docstring and the sorting of the static_data dictionary but for now I would prefer to keep the two helper functions, if that is ok with you.